src/core/finalize.ts

```diff
@@ -43,163 +47,280 @@ export function processResult(result: any, scope: ImmerScope) {
 		}
 	} else {
 		// Finalize the base draft.
-		result = finalize(scope, baseDraft, [])
+		result = finalize(scope, baseDraft)
 	}
+
+	maybeFreeze(scope, result, true)
+
 	revokeScope(scope)
 	if (scope.patches_) {
 		scope.patchListener_!(scope.patches_, scope.inversePatches_!)
 	}
 	return result !== NOTHING ? result : undefined
 }
 
-function finalize(rootScope: ImmerScope, value: any, path?: PatchPath) {
+function finalize(rootScope: ImmerScope, value: any) {
 	// Don't recurse in tho recursive data structures
 	if (isFrozen(value)) return value
 
-	const useStrictIteration = rootScope.immer_.shouldUseStrictIteration()
-
 	const state: ImmerState = value[DRAFT_STATE]
-	// A plain object, might need freezing, might contain drafts
 	if (!state) {
-		each(
-			value,
-			(key, childValue) =>
-				finalizeProperty(rootScope, state, value, key, childValue, path),
-			useStrictIteration
-		)
+		const finalValue = handleValue(value, rootScope.handledSet_, rootScope)
+		return finalValue
+	}
+
+	// Never finalize drafts owned by another scope
+	if (!isSameScope(state, rootScope)) {
 		return value
 	}
-	// Never finalize drafts owned by another scope.
-	if (state.scope_ !== rootScope) return value
+
 	// Unmodified draft, return the (frozen) original
 	if (!state.modified_) {
-		maybeFreeze(rootScope, state.base_, true)
 		return state.base_
 	}
-	// Not finalized yet, let's do that now
+
 	if (!state.finalized_) {
-		state.finalized_ = true
-		state.scope_.unfinalizedDrafts_--
-		const result = state.copy_
-		// Finalize all children of the copy
-		// For sets we clone before iterating, otherwise we can get in endless loop due to modifying during iteration, see #628
-		// To preserve insertion order in all cases we then clear the set
-		// And we let finalizeProperty know it needs to re-add non-draft children back to the target
-		let resultEach = result
-		let isSet = false
-		if (state.type_ === ArchType.Set) {
-			resultEach = new Set(result)
-			result.clear()
-			isSet = true
+		// Execute all registered draft finalization callbacks
+		if (state.callbacks_) {
+			while (state.callbacks_.length > 0) {
+				const callback = state.callbacks_.pop()!
```

Comment:
I'm wondering if you can avoid the `pop` here. Not sure if `.length--` is faster? Or alternative, not modify the callbacks array at all, but separately keep a counter how for we got.